Add Prefer: return=minimal support

garrensmith opened the pull request:
## Overview
Adding this header allows the client to request that only Non-Essential
or All headers are removed from the response. This will decrease the
size of the response.

The idea is to remove all headers that a non-browser client would not need. 
I've had to leave `Server` header in otherwise Mochiweb adds its very lame/snarky header in its place.

There are two options `Non-Essential` which removes everything except caching headers. And then `All` which removes everything except Content-length, Content-Type, Server, Date.

## Testing recommendations

curl -v --header "X-Couch-Exclude-Headers: Non-Essential" 'http://dev:5984/_all_dbs

## GitHub issue number
This PR will be the number


## Checklist

- [x] Code is written and works correctly;
- [x] Changes are covered by tests;
- [ ] Documentation reflects the changes;

rnewson commented on src/chttpd/src/chttpd_exclude_headers.erl:
```diff
@@ -0,0 +1,45 @@
+
+
+% List of essential headers for an http request. If we leave out server then Mochiweb adds in its default one
+-define(ESSENTIAL_HEADERS, ["Cache-Control", "ETag", "Content-Type", "Content-Length", "Vary", "Server"]).
```
list these in alphabetical order.

rnewson replied:
and probably best to lay them out one line per header, so we can modify it over time.

rnewson replied:
perhaps it should even come from config?

rnewson commented:
forgot to add, the structure of the change looks good to me. My comments are mainly about configurability. I like the use of a custom header to modify the response, rather than a query parameter. The recommendation these days is _not_ to add `X-` but there's so much precedence in CouchDB that it might be weird not to always have it. I do think the request header, and its permitted values, are a bit wordy at the same time as being a bit unclear. Perhaps if the filtering were controlled by config this problem goes away.

garrensmith commented:
Thanks @rnewson I've changed it to support using a config setting.

rnewson commented on rel/overlay/etc/default.ini:
```diff
@@ -187,6 +187,12 @@ credentials = false
+; List of headers that will be kept when the header option X-Couch-Exclude-Headers is included in a request
+; 'X-Couch-Exclude-Headers' : "All"
+all = Server
+minimal = Cache-Control, Content-Length, Content-Type, ETag, Server, Vary
```
will need Transfer-Encoding and perhaps more

garrensmith replied:
Good point. I've added Content-Length, Content-Type and Transfer-Encoding to the all list

rnewson commented:
Take a look at RFC 7240 (https://tools.ietf.org/html/rfc7240).

`Prefer: return=minimal`

davisp commented on src/chttpd/src/chttpd_prefer_header.erl:
```diff
@@ -0,0 +1,49 @@
+-module(chttpd_prefer_header).
+
+
+
```
Two blank lines between sections and functions please.

davisp commented on src/chttpd/src/chttpd_prefer_header.erl:
```diff
@@ -0,0 +1,49 @@
+
+
+
+-export([maybe_return_minimal/2]).
```
Exports should go one per line so diff's in the future are nicer.

```erlang
-export([
    maybe_return_minimal/2
]).
```

davisp commented on src/chttpd/test/chttpd_prefer_header_test.erl:
```diff
@@ -0,0 +1,96 @@
+
+
+default_no_exclude_header_test() ->
+    Headers = chttpd_prefer_header:maybe_return_minimal(mock_request([]), default_headers()),
```
Line length too long.

davisp commented on src/chttpd/test/chttpd_prefer_header_test.erl:
```diff
@@ -0,0 +1,96 @@
+-include_lib("eunit/include/eunit.hrl").
+
+
+
```
Only two blank lines between sections and functions please.

davisp commented on rel/overlay/etc/default.ini:
```diff
@@ -188,6 +188,10 @@ credentials = false
 ; List of accepted methods
 ; methods =
 
+[prefer_header]
```
Bike shed: Is this the best place for this, or would it be better as a key in the chttpd section. Something like "prefer_minimal = List-Of-Headers"

You might also add a comment that a user doesn't specify Server here that Mochiweb will add its own regardless.

davisp commented on src/chttpd/src/chttpd_prefer_header.erl:
```diff
@@ -0,0 +1,49 @@
+
+filter_headers(Headers, IncludeList) ->
+    lists:filter(fun({HeaderName, _}) -> 
+        lists:member(HeaderName, IncludeList)
```
Seems like you probably want to look into mochiweb's header handling to make this comparison case insensitive.

davisp replied:
Because users may or may not give headers in the same case as we have them specified internally.

davisp replied:
Also that'd make for a good unit test as well.

rnewson commented on rel/overlay/etc/default.ini:
```diff
@@ -57,6 +57,9 @@ backlog = 512
 require_valid_user = false
+; List of headers that will be kept when the header Prefer: return=minimal is included in a request.
+; If Server header is left out, Mochiweb will add its own one in.
+prefer_minimal = Cache-Control, Content-Length, Content-Type, ETag, Server, Vary
```
we'll need `Transfer-Encoding` otherwise views, changes, _all_dbs will be broken.
we'll need `Set-Cookie` so we can login.
what about CORS related headers?
we'll need `Content-Range` for range reqs on attachments.
we'll need `WWW-Authenticate` for users that enable the standard login prompting setting.

davisp replied:
Agree on Transfer-Encoding and Content-Range. For cooking and authentication headers I think not since this is aimed at non-browser clients and those are both for browsers (and or, non browsers would just have to be smart enough to not request it when they need to get a cookie.

garrensmith replied:
I don't think CORS is required since this is for non-browser clients. I agree with Paul around needing the auth headers, don't use this header when you need a cookie.

rnewson commented on src/chttpd/src/chttpd_prefer_header.erl:
```diff
@@ -0,0 +1,54 @@
+
+-export([
+    maybe_return_minimal/2
+    ]).
```
this line should be at column 0.

garrensmith commented:
Thanks @rnewson and @davisp for the reviews on this. I've updated the `default.ini` and added in the two missing headers. I've also fixed the spacing issue.

rnewson commented on rel/overlay/etc/default.ini:
```diff
@@ -57,6 +57,9 @@ backlog = 512
 require_valid_user = false
+; List of headers that will be kept when the header Prefer: return=minimal is included in a request.
+; If Server header is left out, Mochiweb will add its own one in.
+prefer_minimal = Cache-Control, Content-Length, Content-Range, Content-Type, ETag, Server, Transfer-Encoding, Vary
```
should be commented out by default.

rnewson commented on src/chttpd/src/chttpd_prefer_header.erl:
```diff
@@ -0,0 +1,54 @@
+
+
+get_header_list() ->
+    SectionStr = config:get("chttpd", "prefer_minimal", []),
```
the default needs to be ""?

garrensmith replied:
I've changed it to `""`. It seems` re:split("", "\\s*,\\s*", [trim, {return, list}]).` and ` re:split([], "\\s*,\\s*", [trim, {return, list}]).` return the same result of `[[]]`

rnewson commented on rel/overlay/etc/default.ini:
```diff
@@ -57,6 +57,9 @@ backlog = 512
 require_valid_user = false
+; List of headers that will be kept when the header Prefer: return=minimal is included in a request.
+; If Server header is left out, Mochiweb will add its own one in.
+; prefer_minimal = Cache-Control, Content-Length, Content-Range, Content-Type, ETag, Server, Transfer-Encoding, Vary
```
I'm an idiot. I forgot that this is opt-in (you have to pass 'Prefer: return=minimal'). So please uncomment this again. :)

rnewson commented:
+1 to merge but please squash down from the 7 commits as appropriate.

garrensmith commented:
Docs https://github.com/apache/couchdb-documentation/pull/157
